feat(client): ✨ Appropriate rerouting for game page based on auth

gchamb opened the pull request:
- changed the share url based on josh's suggestion
- Componetized the SigninButton
- reroutes users who are anon and arent in this lobby
- implemented the auth store

gchamb committed:
feat(client): ✨ Start on Joining Lobbies

- Functionality on joining lobbies
- Optional props for Modal
- fixed names

#29

Co-authored-by: mrjoshua520 <joshua52099@gmail.com>

gchamb committed:
refactor(client): ♻️ Error handling for nonexistent lobbies

- simple error handling for if the user enters a code for a lobby that doesn't exist.

gchamb committed:
Merge branch 'development' into gchambers-jsalkeld/join-lobby

gchamb committed:
Merge branch 'development' into gchambers-jsalkeld/join-lobby

gchamb committed:
feat(client): ✨ Subscribing the lobby

- created props for lobby component to pass lobby data
- subscribed the current lobby to its document
- skeleton to changing game state

gchamb committed:
feat: :sparkles: Remove slug url to query parameter

- Able to update game state
- Display name field if anon on join page and signin on join page
- Remove slug to query parameter
- Remove anon popup replaced with above

Co-authored-by: Joshua Salkeld <mrjoshua520@users.noreply.github.com>

gchamb committed:
feat(client): ✨ Able to join off link

- Sends anon users back to join lobby so they can enter their name
- Users can join lobby directly off link if they're not in the lobby already
- adding awaits to functions and lowercasing them

mrjoshua520 committed:
fix(client): 🐛 Fix code displaying as undefined

Code will no longer display as undefined on the join lobby page

gchamb committed:
Merge branch 'development' into gchambers-jsalkeld/join-lobby

gchamb committed:
Merge branch 'gchambers-jsalkeld/join-lobby' of https://github.com/cat-acronym-team/cat-me-if-you-can into gchambers-jsalkeld/join-lobby

gchamb committed:
feat(client): ✨ Appropriate rerouting for game page based on auth

- changed share url based on josh's suggestion
- Componetized the SigninButton
- reroutes users who are anon and arent in this lobby
- implemented the auth store

nstringham commented on client/src/lib/firebase/game.ts:
```diff
@@ -0,0 +1,12 @@
+/*
+  You can change the game state of the current lobby
+*/
+export const changeGameState = async (lobbyId: string, state: GameState) => {
+  await updateDoc(doc(lobbyCollection, lobbyId), {
+    state,
+  });
+};
```
I feel like for security reasons we might want this to be on the server but we can fix this as part of [#30](https://github.com/cat-acronym-team/cat-me-if-you-can/issues/30)

nstringham commented on client/src/components/Modal.svelte:
```diff
@@ -9,10 +9,20 @@
+  export let closeOnClick = false;
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  export let onClosed = () => {};
 </script>
 
```
the recomended way to resolve this warning is to place a comment in the function to make it obvious that you are doing nothing on purpose.
```ts
export let onClosed = () => {
  // Do nothing by default
};
```

nstringham replied:
this comment was supposed to be on line 13 and 14

nstringham commented on client/src/lib/firebase/game.ts:
```diff
@@ -0,0 +1,12 @@
+/*
+  You can change the game state of the current lobby
+*/
+export const changeGameState = async (lobbyId: string, state: GameState) => {
```
can this be a `export function` instead of a `export const` please

nstringham commented on client/src/lib/firebase/join-lobby.ts:
```diff
@@ -0,0 +1,31 @@
+import type { UserData } from "./firestore-types/users";
+import type { Player } from "./firestore-types/lobby";
+
+export const findAndJoinLobby = async (id: string, allUserInfo: UserData & { uid: string }) => {
```
can this be a `export function` instead of a `export const` please

nstringham commented on client/src/lib/firebase/join-lobby.ts:
```diff
@@ -0,0 +1,31 @@
+        alive: true,
+      },
+    ],
+  });
```
if you want this to be on the client you defintily need to write a security rule for this

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -0,0 +1,104 @@
+        uid: $user.uid,
+      });
+      // go to game page
+      goto(`/game?=${code}`);
```
the url querry should be a key value pair. something like `/game?code=${code}`

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -0,0 +1,104 @@
+  onMount(() => {
+    // gets code from url search
+    // the svelte magic with searchparams wasnt working
+    const queryCode = $page.url.search.split("=")[1];
```
this will fail if the url is something like `/join?src=share-button&code=abcxyz

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -0,0 +1,104 @@
+      });
+      // go to game page
+      goto(`/game?=${code}`);
+    } catch (err: any) {
```
try not to use any.

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -0,0 +1,56 @@
+
+  onMount(async () => {
+    // gets code from url search
+    // the svelte magic with searchparams wasnt working
```
it probobly wasn't working because of the mistake on line 46 of the join page
```ts
goto(`/game?=${code}`);
```

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -0,0 +1,56 @@
+    // the new doc from subscribing below. This makes the redirect slow since we want to check if the user isn't in this lobby
+    const lobbyDoc = await getDoc(doc(lobbyCollection, code));
+    if (!lobbyDoc.exists()) {
+      goto("/join");
```
it might be nice to pass the code to join so that it can display an error message

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -0,0 +1,104 @@
+    <p style="color:red;">{error.message}</p>
+  {/if}
+  <form on:submit|preventDefault={joinLobby}>
+    {#if $user === null}
+      <input type="text" placeholder="Enter in your display name" bind:value={name} />
+    {/if}
```
it might be better to have this textbox always visable so that way if I already have an account I can do a last minute name changebefore I join

nstringham reviewed (changes requested):
seams to be working pretty good just a few little things to address

mrjoshua520 committed:
refactor(client): 🎨 Update various aspects of the join lobby code

-Updated comments in modal.svelte
-Changed export const to export function
-changed url to key value pair
-changed split to look for code= rather than just =
-display name text box always visable

gchamb committed:
feat(client): ✨ Implement requested changes

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -44,11 +69,12 @@
+    } catch (err) {
       // if the lobby doesn't exist then error is thrown
+      const myError = err as Error;
       error = {
```
Is this a safe type assertion? I feel like it might be better to use an if.

gchamb replied:
the err parameter in the catch is either unknown or any so you literally have to tell typescript its type to be able to use it.

nstringham replied:
the reason it is `unknown`  is because we actually don't know what it is. it might not be an error. in JavaScript you can throw any value.

here is what we can do:

at runtime check if it is an error using the JavaScript [instanceof operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof)

if it is an error then you can use it's message

if it is not an error then you still want to display something to the user so we can use the [String Constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/String) without the `new` keyword which can convert any value to a string

we could use an if statement like this:

```ts
catch (err) {
  // if the lobby doesn't exist then error is thrown
  let errorMessage: string;
  if (err instanceof Error) {
    errorMessage = err.message;
  } else {
    errorMessage = String(err);
  }
  error = {
    status: true,
    message: errorMessage,
  };
  code = "";
}
```

or we can do it with the [ternary operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) if you want
```ts
catch (err) {
  // if the lobby doesn't exist then error is thrown
  error = {
    status: true,
    message: err instanceof Error ? err.message : String(err),
  };
  code = "";
}
```

gchamb replied:
Okay that makes sense for the unknown type

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -0,0 +1,128 @@
+    }
+    // gets code from url search
+    // the svelte magic with searchparams wasnt working
+    const queryCode = $page.url.search.split("code=")[1];
```
this will not work with a url that looks like `/join?code=abcxyz&src=share-button`

why not use something like this?
```ts
$page.url.searchParams.get("code")
```

nstringham replied:
this will not work with a url that looks like `/join?code=abcxyz&src=share-button`

why not use something like this?
```ts
$page.url.searchParams.get("code")
```

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -0,0 +1,68 @@
+  onMount(async () => {
+    // gets code from url search
+    // the svelte magic with searchparams wasnt working
+    code = $page.url.search.split("code=")[1];
```
this will not work with a url that looks like `/game?code=abcxyz&src=share-button`

why not use something like this?
```ts
$page.url.searchParams.get("code")
```

nstringham reviewed (changes requested):
if I am already signed in and I click the join button then the display name box is empty instead of showing the Display name I used last time.

gchamb committed:
feat(client): ✨ Finish Requested Changes

- fixed error typing in catch block
- fixed name population on join page
- put the update functions on the server
- added functions to svelte app
- added some types for changing state(since we might be using that often)

gchamb committed:
fix(client): 🐛 prettier issue

gchamb committed:
style(client): 🎨 fix prettier pt2

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -25,22 +32,24 @@
+        on:click={async () => {
+          await startGame({ code, state: "PROMPT" });
+        }}>Start Game</button
+      >
```
the await isn't really doing anything here but it's also not hurting anything so whatever

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -1,15 +1,22 @@
+  // Josh's suggestion that I agreed on
+  let url = `${$page.url.origin}/join?code=${code}`;
   let canShare = false;
+  const startGame = httpsCallable<ChangeState, ChangeStateResponse>(functions, "changeState");
```
might be a good idea to put this into a seprate file kinda like we do for `firestore-collections.ts` but this is fine for now

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -23,6 +23,8 @@ export type Player = {
 
 export type GameState = "WAIT" | "PROMPT" | "CHAT" | "VOTE" | "END";
+export type ChangeState = { code: string; state: GameState };
+export type ChangeStateResponse = { error?: string };
```
this isn't really a "firestore" type maybe it would be a good idea to make a seprate file for firestore functions types?

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+  // return
+  return lobbyCollection.doc(data.code).update({ state: data.state });
+});
```
definitly need some security here we don't want any logged in player to be able to chnage the sate to anytrhing whenever they feel like

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,37 @@
+    players: [...players, { displayName: data.displayName, avatar: data.avatar as Player["avatar"], alive: true }],
+    uids: [...uids, data.uid],
+  });
+});
```
definitly need secrity and validation here as well

nstringham commented on client/src/lib/firebase/join-lobby.ts:
```diff
@@ -0,0 +1,28 @@
+    throw new Error("You are already in the lobby!");
+  }
+  // make request to server
+  const { data } = await httpsCallable<UserData & { uid: string; code: string }, { error?: string }>(
```
probobly a good idea to put this into a seprate file kinda like we do for `firestore-collections.ts`

nstringham commented on client/src/lib/firebase/join-lobby.ts:
```diff
@@ -0,0 +1,28 @@
+    "addPlayer"
+  )({ ...allUserInfo, code: id });
+  if (data.error !== undefined) {
+    throw data.error;
```
should probobly wrap this in an error

nstringham commented on client/src/lib/firebase/splash.ts:
```diff
@@ -1,6 +1,36 @@
+  3. A User with User Doc
+    - just update their display name
+*/
+export const saveOrCreate = async (user: User | null, userData: UserData | undefined, name: string) => {
```
can we do `export function` instead of `export const` please

gchamb committed:
Merge branch 'development' into gchambers-jsalkeld/join-lobby

gchamb committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into gchambers-jsalkeld/join-lobby

gchamb committed:
feat(functions): 🔒️ Add Security to HTTP Callables

- implemented request for better file structure
- removed client data from callables except the code
- secured the callables

gchamb committed:
style(functions): 🎨 format with prettier

nstringham commented on client/src/lib/firebase/firestore-functions-types.ts:
```diff
@@ -0,0 +1,4 @@
+import type { GameState } from "./firestore-types/lobby";
+
+export type ChangeState = { code: string; state: GameState };
+export type ErrorResponse = { error?: string };
```
we should probably share types between the client and the server like we did for firestore-types

nstringham commented on client/src/lib/firebase/firestore-functions-types.ts:
```diff
@@ -0,0 +1,4 @@
+import type { GameState } from "./firestore-types/lobby";
+
+export type ChangeState = { code: string; state: GameState };
```
is this used anywhere anymore?

nstringham commented on client/src/lib/firebase/firestore-functions.ts:
```diff
@@ -0,0 +1,6 @@
+import { httpsCallable } from "firebase/functions";
+import type { ErrorResponse } from "./firestore-functions-types";
+
+export const startGame = httpsCallable<{ code: string }, ErrorResponse>(functions, "startGame");
+export const addPlayer = httpsCallable<{ code: string }, ErrorResponse>(functions, "addPlayer");
```
we should probobly makes types for the imnput data and share the types between the server and the client so that if we want to change it in the future it will force us to change it in both places

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,56 @@
+import { Lobby } from "./firestore-types/lobby";
+import { UserData } from "./firestore-types/users";
+
+export const startGame = functions.https.onCall(async (data: { code: string }, context) => {
```
we should probobly makes types for the imnput data and share the types between the server and the client so that if we want to change it in the future it will force us to change it in both places

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,56 @@
+  return lobbyCollection.doc(data.code).update({ state: "PROMPT" });
+});
+
+export const addPlayer = functions.https.onCall(async (data: { code: string }, context) => {
```
we should probobly makes types for the imnput data and share the types between the server and the client so that if we want to change it in the future it will force us to change it in both places

nstringham replied:
maybe we should call this joinLobby now that the uid is not in the input data?

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,56 @@
+    return { error: "Not Signed In" };
+  }
+  // get lobby doc
+  const lobby = await lobbyCollection.doc(data.code).get();
```
we don't know that data is an object where code is a string so we should do some validation here.

here is how me and nathan are doing it for avatar a2ff3094d4203331f0dea0ee3f5012b46b5ee47a

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,56 @@
+    return { error: "Not Signed In" };
+  }
+  // lobby doc
+  const lobby = lobbyCollection.doc(data.code);
```
we don't know that data is an object where code is a string so we should do some validation here.

here is how me and nathan are doing it for avatar a2ff3094d4203331f0dea0ee3f5012b46b5ee47a

gchamb committed:
refactor: ♻️ Share the types

- share the types of client and server
- switch name
- validate code

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -23,7 +23,6 @@ export type Player = {
 };
 
 export type GameState = "WAIT" | "PROMPT" | "CHAT" | "VOTE" | "END";
-
```
lets keep this empty line

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,65 @@
+    return { error: "Not Signed In" };
+  }
+  // validate code
+  if (codeCheck(data.code) === false) {
```
this asumes that data is an object but it might be something else

nstringham commented on functions/src/lobby.ts:
```diff
@@ -0,0 +1,65 @@
+    return { error: "Not Signed In" };
+  }
+  // validate code
+  if (codeCheck(data.code) === false) {
```
this asumes that data is an object but it might be something else

gchamb committed:
fix(functions): 🐛 Validate the data parameter

nstringham committed:
refactor(functions): ♻️ use `!` instead of `=== false` for type predicates

I don't know why `=== false` doeasn't work but yeah

nstringham committed:
refactor(functions): ♻️ rename codeCheck to isLobbyRequest

to improve readablity

nstringham committed:
refactor(functions): ⚰️ remove unused import

nstringham committed:
test(functions): ✅ add tests for isLobbyRequest

nstringham committed:
fix(functions): 🐛 catch if lobby request is undifined or null

nstringham committed:
refactor(functions): ♻️ use `typeof` and `in` to avoid `as` in `isLobbyRequest`

this is a more idomatic way of checking

nstringham reviewed (approved):
just a few small issues so I fixed them myself in the interest of time
